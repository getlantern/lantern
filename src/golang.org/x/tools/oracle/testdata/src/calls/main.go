package main

// Tests of call-graph queries.
// See go.tools/oracle/oracle_test.go for explanation.
// See calls.golden for expected query results.

func A(x *int) { // @pointsto pointsto-A-x "x"
	// @callers callers-A "^"
	// @callstack callstack-A "^"
}

func B(x *int) { // @pointsto pointsto-B-x "x"
	// @callers callers-B "^"
}

// apply is not (yet) treated context-sensitively.
func apply(f func(x *int), x *int) {
	f(x) // @callees callees-apply "f"
	// @callers callers-apply "^"
}

// store *is* treated context-sensitively,
// so the points-to sets for pc, pd are precise.
func store(ptr **int, value *int) {
	*ptr = value
	// @callers callers-store "^"
}

func call(f func() *int) {
	// Result points to anon function.
	f() // @pointsto pointsto-result-f "f"

	// Target of call is anon function.
	f() // @callees callees-main.call-f "f"

	// @callers callers-main.call "^"
}

func main() {
	var a, b int
	apply(A, &a) // @callees callees-main-apply1 "app"
	apply(B, &b)

	var c, d int
	var pc, pd *int // @pointsto pointsto-pc "pc"
	store(&pc, &c)
	store(&pd, &d)
	_ = pd // @pointsto pointsto-pd "pd"

	call(func() *int {
		// We are called twice from main.call
		// @callers callers-main.anon "^"
		return &a
	})

	// Errors
	_ = "no function call here"   // @callees callees-err-no-call "no"
	print("builtin")              // @callees callees-err-builtin "builtin"
	_ = string("type conversion") // @callees callees-err-conversion "str"
	call(nil)                     // @callees callees-err-bad-selection "call\\(nil"
	if false {
		main() // @callees callees-err-deadcode1 "main"
	}
	var nilFunc func()
	nilFunc() // @callees callees-err-nil-func "nilFunc"
	var i interface {
		f()
	}
	i.f() // @callees callees-err-nil-interface "i.f"

	i = new(myint)
	i.f() // @callees callees-not-a-wrapper "f"
}

type myint int

func (myint) f() {
	// @callers callers-not-a-wrapper "^"
}

var dynamic = func() {}

func deadcode() {
	main() // @callees callees-err-deadcode2 "main"
	// @callers callers-err-deadcode "^"
	// @callstack callstack-err-deadcode "^"

	// Within dead code, dynamic calls have no callees.
	dynamic() // @callees callees-err-deadcode3 "dynamic"
}

// This code belongs to init.
var global = 123 // @callers callers-global "global"

// The package initializer may be called by other packages' inits, or
// in this case, the root of the callgraph.  The source-level init functions
// are in turn called by it.
func init() {
	// @callstack callstack-init "^"
}
