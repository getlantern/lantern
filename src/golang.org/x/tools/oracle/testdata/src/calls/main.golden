-------- @pointsto pointsto-A-x --------
this *int may point to these objects:
	a
	b

-------- @callstack callstack-A --------
Found a call path from root to main.A
main.A
dynamic function call from main.apply
static function call from main.main

-------- @pointsto pointsto-B-x --------
this *int may point to these objects:
	a
	b

-------- @callers callers-B --------
main.B is called from these 1 sites:
	dynamic function call from main.apply

-------- @callees callees-apply --------
this dynamic function call dispatches to:
	main.A
	main.B

-------- @callers callers-apply --------
main.apply is called from these 2 sites:
	static function call from main.main
	static function call from main.main

-------- @callers callers-store --------
main.store is called from these 2 sites:
	static function call from main.main
	static function call from main.main

-------- @pointsto pointsto-result-f --------
this func() *int may point to these objects:
	main.main$1

-------- @callees callees-main.call-f --------
this dynamic function call dispatches to:
	main.main$1

-------- @callers callers-main.call --------
main.call is called from these 2 sites:
	static function call from main.main
	static function call from main.main

-------- @callees callees-main-apply1 --------
this static function call dispatches to:
	main.apply

-------- @pointsto pointsto-pc --------
this *int may point to these objects:
	c

-------- @pointsto pointsto-pd --------
this *int may point to these objects:
	d

-------- @callees callees-err-no-call --------

Error: there is no function call here
-------- @callees callees-err-builtin --------

Error: this is a call to the built-in 'print' operator
-------- @callees callees-err-conversion --------

Error: this is a type conversion, not a function call
-------- @callees callees-err-bad-selection --------

Error: ambiguous selection within function call (or conversion)
-------- @callees callees-err-deadcode1 --------
this static function call dispatches to:
	main.main

-------- @callees callees-err-nil-func --------
dynamic function call on nil value

-------- @callees callees-err-nil-interface --------
dynamic method call on nil value

-------- @callees callees-not-a-wrapper --------
this dynamic method call dispatches to:
	(main.myint).f

-------- @callers callers-not-a-wrapper --------
(main.myint).f is called from these 1 sites:
	dynamic method call from main.main

-------- @callees callees-err-deadcode2 --------
this static function call dispatches to:
	main.main

-------- @callstack callstack-err-deadcode --------
main.deadcode is unreachable in this analysis scope

-------- @callees callees-err-deadcode3 --------

Error: this call site is unreachable in this analysis
-------- @callers callers-global --------
main.init is called from these 1 sites:
the root of the call graph

-------- @callstack callstack-init --------
Found a call path from root to main.init#1
main.init#1
static function call from main.init

